### ISSUE TITLE
Are any of these functions worth adding here?

### ISSUE THREAD

**Issue by NeilGirdhar** [data-apis/array-api-extra#139]
I'm just wondering if any of these functions are worth adding to this project:

https://github.com/NeilGirdhar/tjax/blob/main/tjax/_src/math_tools.py

https://github.com/NeilGirdhar/tjax/blob/main/tjax/_src/leaky_integral.py

The leaky integral would need a lax.scan, which was deferred by the Array API people.

**Comment by NeilGirdhar** [comment]
> The functionality of `divide_where` is delivered by [#14](https://github.com/data-apis/array-api-extra/issues/14).

I don't think so.   While both functions can produce the same primals, `divide_where` prevents NaN on the cotangents.

> `leaky_integrate` sounds like something backed by some paper?

Leaky integration is extremely common in machine learning.  It's a special case of [`lfilter`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html).  But to implement `lfilter` in the Array API, you need `lax.scan` or something like it.

**Comment by crusaderky** [comment]
> I don't think so. While both functions can produce the same primals, `divide_where` prevents NaN on the cotangents.

Could you produce an example where the snippet below doesn't replicate the functionality of `divide_where`?
```python
def divide_where(dividend, divisor, where, otherwise):
    return _lazywhere(
        where, 
        lambda n, d, o: n / d, 
        dividend, divisor, otherwise, 
        f2=lambda n, d, o: o,
    )
```

> It's a special case of [lfilter](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html).

To me `lax.scan` sounds like a useful primitive with a place in array-api-extra; `leaky_integrate` definitely doesn't.
I'm unsure however on how an xpx implementation of `lax.scan` would look like for backends other than JAX. cython for numpy (which would be a major problem for vendoring) and iterate cell-by-cell in pure python for all other backends? That could be very slow depending on your array's size along the target axis.

**Comment by NeilGirdhar** [comment]
> Could you produce an example where the snippet below doesn't replicate the functionality of `divide_where`?

As I said above: "While both functions can produce the same primals, divide_where prevents NaN on the cotangents."  For example,

```python
def divide_where_b(dividend: Any, divisor: Any, where: Any, otherwise: Any) -> Any:
    return apply_where(  # Had to use apply_where since _lazywhere doesn't support Jax arrays.
        where,
        lambda n, d, o: n / d,
        lambda n, d, o: o,
        dividend, divisor, otherwise,
    )


x = jnp.zeros(())


def f(x: Any) -> Any:
    return divide_where(x, x, where=x != 0.0, otherwise=jnp.ones(()))


def g(x: Any) -> Any:
    return divide_where_b(x, x, where=x != 0.0, otherwise=jnp.ones(()))


a = grad(f)(x)
b = grad(g)(x)
print(a, b)  # 0.0 nan
```

> To me `lax.scan` sounds like a useful primitive with a place in array-api-extra;

I agree.  Ideally that `scan` could be used to efficiently implement SciPy's filters.

**Comment by NeilGirdhar** [comment]
If there's consensus, should I implement `scan`?  I imagine it's just a for-loop except for Jax arrays where it delegates to `lax.scan`?

**Comment by crusaderky** [comment]
Right. So there is no difference unless you run inside an auto-differentiation tool, which AFAIK is exclusively offered by JAX, at least among array API compliant libraries. ndonnx could potentially offer one too in the future I guess. Dask _could_ as a graph-based engine but such a feature would be firmly out of scope for the project.

Here's the problem: let's assume that array-api-extra offers both `apply_where` and `divide_where`. The only way to test that a downstream library (read: scipy) is using one or the other is with `jax.grad`. Would you expect all scipy functions to eventually gain a jax-only test that verifies their output when they're wrapped by jax.grad?

**Comment by rgommers** [comment]
> an auto-differentiation tool, which AFAIK is exclusively offered by JAX,

And PyTorch, and MLX.

That said, I agree - auto-differentiation isn't in scope, that's whole new level of complexity. Seems like a bridge too far to do in a framework-agnostic way at the moment.

**Comment by NeilGirdhar** [comment]
> Right. So there is no difference unless you run inside an auto-differentiation tool, which AFAIK is exclusively offered by JAX, at least among array API compliant libraries.

Do the other Array API libraries not offer differentiation?  Pytorch and Tensorflow do.

I also disagree that the effect on cotangents doesn't matter.  The effect on primals and cotangents are part of the API.  If you implement something in a way that destroys or damages the cotangents, then that is a different API.  Care should be taken to always produce the right ones since they can have dramatic effects on algorithms that use the functions.

> Would you expect all scipy functions to eventually gain a jax-only test that verifies their output when they're wrapped by jax.grad?

It's not Jax-only, and yes, I think this library should be testing that the cotangents aren't damaged because it will break _normal uses of this library if the cotangents are damaged._

**Comment by lucascolley** [comment]
thanks for thinking about this @NeilGirdhar, I'll close this as 'not planned' for now with the autodiff label. It sounds like we should return to this in the future if there is wider work on autodiff within the standard ecosystem, but this issue is maybe not the best place for that to start.

**Comment by NeilGirdhar** [comment]
Yeah, no worries, I think I got a bit carried away thinking of all the bugs I've worked through tracing cotangents.  But I'll just keep any related tools in tjax for now.

However, as more and more algorithms get ported to the Array API, I imagine the `stop_gradient` will be requested pretty early on:
```python
def stop_gradient(x: U, *, xp: ModuleType | None = None) -> U:
    if xp is None:
        xp = get_namespace(x)
    if is_jax_array(xp):
        from jax.lax import stop_gradient as sg  # noqa: PLC0415
        return sg(x)
    if is_torch_array(xp):
        from torch import Tensor  # noqa: PLC0415
        assert isinstance(x, Tensor)
        return x.detach()  # pyright: ignore
    return x
```
No rush, of course.
